plugins/state_history_plugin/state_history_plugin.cpp

```diff
@@ -292,7 +292,7 @@ struct state_history_plugin_impl : std::enable_shared_from_this<state_history_pl
              "the process");
       }
 
-      boost::asio::post(get_ship_executor(), [self = this->shared_from_this(), block_state]() {
+      boost::asio::post(get_ship_executor(), [self = this->shared_from_this(), &block_state]() {
```

Comment:
Does this cause a race where this `block_state` can be destroyed before this lambda is executed? Is there something else that can guarantee `block_state`'s ref count remains alive?

Reply:
If this is the memory "leak", then we just need to apply back pressure.

Reply:
I think you might just be able to skip this post() if the ship thread pool hasn't been started? The problem is replay happens synchronously in chain_plugin's initialize -- before the ship threads are started -- leading to all theses post()s just stacking up until replay is complete.

Since there won't be any ship clients connected until after replay is complete, delaying these posts until the threads are running seems reasonable..

Reply:
wait.. that can't be right.. since the signal isn't connected until after state_history plugin's initaialize either, hm..

Reply:
Replay happens in chain_plugin plugin_startup. We can just skip the post if there are no sessions which is affectively what was done before.

Reply:
ah okay, then yeah makes sense

Reply:
Tried that and it works, that was definitely it :) Thanks a lot @spoonincode and @heifner!